### PR TITLE
Build with VS2015

### PR DESCRIPTION
**Pull request by jeffhostetler**
This sequence of commits allows git-for-windows to be built using VS2015.
This includes changes to build scripts and MSVC-specific versions of various OS and CRT wrappers.
It is sufficient to allow debugging and profiling using VS2015.

It DOES NOT attempt to generate SLN or VCPROJ files.
Most of my efforts here have been focused on porting to VC and the new UCRT in VS2015.

I'd eventually like to merge this effort with: https://github.com/git-for-windows/git/pull/256
since Philip is concentrating on the SLN and VCPROJ issues.

This should be considered a WIP, since there are several cleanup steps (documented in one
of the READMEs) and I still need to get through the test suite (and add any additional tests we
think are necessary).

### CODE REVIEW

**Comment by PhilipOakley** [comment]
----- Original Message ----- 
  From: Jeff Hostetler 
  To: git-for-windows/git 
  Sent: Friday, June 03, 2016 9:22 PM
  Subject: [git-for-windows/git] WIP -- Build with VS2015 (#773)


  This sequence of commits allows git-for-windows to be built using VS2015.
  This includes changes to build scripts and MSVC-specific versions of various OS and CRT wrappers.
  It is sufficient to allow debugging and profiling using VS2015.

  It DOES NOT attempt to generate SLN or VCPROJ files.
  Most of my efforts here have been focused on porting to VC and the new UCRT in VS2015.

  I'd eventually like to merge this effort with: #256
  since Philip is concentrating on the SLN and VCPROJ issues.

Excellent work
I suspect that the #256 will work on top of this

Philip

------------------------------------------------------------------------------

  You can view, comment on, or merge this pull request online at:
    https://github.com/git-for-windows/git/pull/773

  Commit Summary
    a.. Add VS2015 trash files to .gitignore 
    b.. NuGet scripts for building with VS2015 
    c.. Update compat/vcbuild/scripts/clink.pl for VS2015 
    d.. Updated Makefile and compiler settings for VS2015. 
    e.. msvc.h: include sigset_t definition 
    f.. compat/msvc.h: define O_ACCMODE for MSVC 
    g.. _REPARSE_DATA_BUFFER: add MSCV definitions 
    h.. fixup! _REPARSE_DATA_BUFFER: 
    i.. Compile with VS2015. 
    j.. Turned off debug messages in cache-tree.c 
  File Changes
    a.. M .gitignore (2) 
    b.. M Makefile (2) 
    c.. M cache-tree.c (10) 
    d.. M compat/mingw.c (342) 
    e.. M compat/mingw.h (63) 
    f.. M compat/msvc.h (40) 
    g.. A compat/vcbuild/.gitignore (1) 
    h.. A compat/vcbuild/Makefile (124) 
    i.. A compat/vcbuild/README_VS2015.txt (62) 
    j.. A compat/vcbuild/nuget.config (27) 
    k.. A compat/vcbuild/packages.config (23) 
    l.. M compat/vcbuild/scripts/clink.pl (10) 
    m.. M compat/winansi.c (114) 
    n.. M config.mak.uname (42) 
    o.. M git-compat-util.h (9) 
  Patch Links:
    a.. https://github.com/git-for-windows/git/pull/773.patch 
    b.. https://github.com/git-for-windows/git/pull/773.diff

**Comment by jeffhostetler** [comment]
Rebased onto v.2.9.0.windows.1.

File: compat/vcbuild/scripts/clink.pl [hunk 1]
```diff
@@ -12,10 +12,11 @@
 use strict;
 my @args = ();
 my @cflags = ();
+my @lflags = ();
 my $is_linking = 0;
 while (@ARGV) {
 	my $arg = shift @ARGV;
-	if ("$arg" =~ /^-[DIMGO]/) {
+	if ("$arg" =~ /^-[DIMGOZ]/) {
```
Comment (dscho) [inline compat/vcbuild/scripts/clink.pl]: What is `-Z` supposed to do? Could you describe that in the commit message, and if necessary, add an argument why we simply pass it through?

File: Makefile [hunk 2]
```diff
@@ -987,7 +987,7 @@ endif
 
 ifdef SANE_TOOL_PATH
 SANE_TOOL_PATH_SQ = $(subst ','\'',$(SANE_TOOL_PATH))
-BROKEN_PATH_FIX = 's|^\# @@BROKEN_PATH_FIX@@$$|git_broken_path_fix $(SANE_TOOL_PATH_SQ)|'
+BROKEN_PATH_FIX = "s|^\# @@BROKEN_PATH_FIX@@$$|git_broken_path_fix '$(SANE_TOOL_PATH_SQ)'|"
```
Comment (dscho) [inline Makefile]: We probably should use `\\#` instead of `\#` now, too (even if it seems to be accepted by Bash).

File: config.mak.uname [hunk 3]
```diff
@@ -335,6 +335,16 @@ endif
...
 	GIT_VERSION := $(GIT_VERSION).MSVC
 	pathsep = ;
+	# Prepend MSVC 64-bit tool-chain to PATH.
+	#
+	# TODO A normal git-bash window DOES NOT contain the path to CL.EXE
+	# TODO and friends in $PATH.
+	# TODO
+	# TODO Warning: There is a MS link.exe next to CL.exe and a gcc link.exe
+	# TODO next to gcc.exe, so we have to prepend this onto the existing
+	# TODO $PATH.  (Or fix clink.pl and lib.pl to use full paths.)
```
Comment (dscho) [inline config.mak.uname]: Prepending is the right thing to do, I am convinced by this argument. Let's skip the part included in the parentheses.

File: config.mak.uname [hunk 4]
```diff
@@ -335,6 +335,16 @@ endif
 ifeq ($(uname_S),Windows)
 	GIT_VERSION := $(GIT_VERSION).MSVC
 	pathsep = ;
+	# Prepend MSVC 64-bit tool-chain to PATH.
+	#
+	# TODO A normal git-bash window DOES NOT contain the path to CL.EXE
+	# TODO and friends in $PATH.
```
Comment (dscho) [inline config.mak.uname]: I think we can probably auto-detect the latest `vcvars.bat` using a wildcard expression. Will try to get that to work later this week.

File: compat/mingw.c [hunk 5]
```diff
@@ -2303,7 +2303,7 @@ int symlink(const char *target, const char *link)
 	return 0;
 }
 
-#ifndef _WINNT_H
+#if !defined(_WINNT_H) && !defined(_MSC_VER)
```
Comment (dscho) [inline compat/mingw.c]: There's a tyop in the commit message: it is "ANONYMOUS" with an "M"... ;-)

Comment (dscho) [inline compat/mingw.c]: Oh, and: would it not be better to simply guard the `_ANONYMOUS_UNION` in an `#ifndef _MSC_VER ... #endif`?

File: compat/mingw.c [hunk 6]
```diff
@@ -1418,7 +1570,11 @@ static int try_shell_exec(const char *cmd, char *const *argv)
 	prog = path_lookup(interpr, path, 1);
 	if (prog) {
 		int argc = 0;
+#if defined(_MSC_VER)
+		char **argv2;
+#else
 		const char **argv2;
+#endif
```
Comment (dscho) [inline compat/mingw.c]: I would prefer:

```c
#ifndef _MSC_VER
        const
#endif
        char **argv2;
```

would you agree?

File: compat/mingw.h [hunk 7]
```diff
@@ -344,11 +385,13 @@ static inline long long filetime_to_hnsec(const FILETIME *ft)
 #ifndef __MINGW64_VERSION_MAJOR
 #define off_t off64_t
 #define lseek _lseeki64
+#ifndef _MSC_VER
```
Comment (dscho) [inline compat/mingw.h]: In the interest of cherry-pick'ability, could you extract this guard into its own commit?

File: compat/vcbuild/README_VS2015.txt [hunk 8]
```diff
@@ -9,15 +9,54 @@ Installing third-party dependencies:
     https://dist.nuget.org/index.html
 
 [2] Download required nuget packages for third-party libraries.
-    Using a bash shell window, type:
+    Using a BASH SHELL window, type:
```
Comment (dscho) [inline compat/vcbuild/README_VS2015.txt]: Personally, I think both are wrong ;-) Could we back out this change?

File: compat/winansi.c [hunk 9]
```diff
@@ -8,6 +8,22 @@
 #include <winreg.h>
 #include "win32.h"
 
+#if defined(_MSC_VER)
```
Comment (dscho) [inline compat/winansi.c]: It would be really, really awesome if these changes could be moved into their own commit. I really think that the improved `isatty()` handling deserves its own patch.

File: config.mak.uname [hunk 10]
```diff
@@ -362,13 +362,7 @@ ifeq ($(uname_S),Windows)
...
 	NO_MKDTEMP = YesPlease
 	NO_MKSTEMPS = YesPlease
-	# VS2015 with UCRT claims that snprintf and friends are C99 compliant,
-	# so we don't need this.
-	#
-	# TODO If we want to support older compilers, we need to make this
-	# TODO conditional on the compiler version.
-	#
-	# SNPRINTF_RETURNS_BOGUS = YesPlease
+	SNPRINTF_RETURNS_BOGUS = YesPlease
```
Comment (dscho) [inline config.mak.uname]: Oversight? Or is it true that `snprintf()` returns a bogus value?

File: cache-tree.c [hunk 11]
```diff
@@ -110,7 +110,7 @@ static int do_invalidate_path(struct cache_tree *it, const char *path)
 	int namelen;
 	struct cache_tree_sub *down;
 
-#if DEBUG
+#if 0 && DEBUG
```
Comment (dscho) [inline cache-tree.c]: I am pretty certain that the `DEBUG` constant does not have the same meaning in `gcc` as it has in MSVC. As such, I believe that the original author meant for this constant to be `#define`d by a developer within the file (see 749864627c2d3c33bbc56d7ba0b5542af698cc40 for the original contribution that seems to concur with my assessment).

Therefore, I think the best course of action would be to rename all instances of `DEBUG` to `DEBUG_CACHE_TREE` in this file.

File: usage.c [hunk 12]
```diff
@@ -15,7 +15,12 @@ void vreportf(const char *prefix, const char *err, va_list params)
 
 	fflush(fh);
 	if (!tweaked_error_buffering) {
+#if defined(_MSC_VER)
+		/* UCRT doesn't like zero buffer size */
+		setvbuf(fh, NULL, _IOLBF, BUFSIZ);
+#else
 		setvbuf(fh, NULL, _IOLBF, 0);
+#endif
```
Comment (dscho) [inline usage.c]: Couldn't we simply replace the entire call with `setlinebuf(fh);`?

Comment (dscho) [inline usage.c]: (I just found out that it is a BSD compatibility function, so maybe it is not such a good idea...)

Comment (dscho) [inline usage.c]: My preference would be to replace the call for all platforms, not guard it. If it works, of course... (it worked in a Linux VM here, so...)

File: Makefile [hunk 13]
```diff
@@ -2332,6 +2332,28 @@ install: all
...
+	$(INSTALL) git-daemon.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-fast-import.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-http-backend.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-http-fetch.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-http-push.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-imap-send.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-remote-http.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-remote-testsvn.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-sh-i18n--envsubst.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
+	$(INSTALL) git-show-index.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
```
Comment (dscho) [inline Makefile]: Maybe something like:

```make
$(INSTALL) $(patsubst %,%.pdb,$(BINDIR_PROGRAMS_NEED_X)) \
        vc*.pdb '$(DESTDIR_SQ)$(bindir_SQ)'
$(INSTALL) *.pdb '$(DESTDIR_SQ)$(gitexec_instdir_SQ)'
```

instead?

**Comment by dscho** [comment]
Looks quite good already!

I am sure we can whip this into shape now.

**Comment by jeffhostetler** [comment]
Hey @dscho big thanks for taking a look at this!  I'll give it another spin shortly.

**Comment by jeffhostetler** [comment]
Note to self: See also #49  WRT the environment conversion issue.  I added code in my port to MSVC to change how that is handled at startup due to differences in the CRT.

File: usage.c (hunk 12 above)
Comment (KindDragon) [inline usage.c]: I think here you should use unconditionally `_IONBF` http://www.cplusplus.com/reference/cstdio/setvbuf/

Comment (dscho) [inline usage.c]: @KindDragon But that would revert https://github.com/jeffhostetler/git/commit/f4c3edc0b156362a92bf9de4f0ec794e90a757fc. In other words, you suggested a regression for platforms other than Windows, at least.

Comment (KindDragon) [inline usage.c]: You mean:
> If buffer set to a null pointer, the function automatically allocates a buffer.

But the documentation does not say that the function supports the zero size, so _IOLBF should be always with non-zero size as you suggesting.

Comment (dscho) [inline usage.c]: No, I meant that `_IONBF` is actively wrong. That is all I said...

Comment (dscho) [inline usage.c]: The commit that you suggested to revert intended to switch to line buffered mode, nothing more, nothing less. [Linux' man page on `setvbuf`](http://linux.die.net/man/3/setvbuf) states:

> If the argument `buf` is `NULL`, only the mode is affected;

In other words, the size *is* ignored if `buf` is set to `NULL`. It gets worse: the man page then goes on to explain the `setlinebuf()` variant (which conforms only to 4.2BSD, hence we cannot use it):

>  The `setlinebuf()` function is exactly equivalent to the call:
> ```c
>     setvbuf(stream, NULL, _IOLBF, 0); 
> ```

Please note that this is precisely what Git's source code reads at the moment.

As you can see, the documentation you are most likely to find on the 'net disagrees with this statement:

> But the documentation does not say that the function supports the zero size, so _IOLBF should be always with non-zero size as you suggesting

Comment (KindDragon) [inline usage.c]: Ok, thanks for explanation.

File: compat/vcbuild/scripts/clink.pl (hunk 1 above)
Comment (jeffhostetler) [inline compat/vcbuild/scripts/clink.pl]: Leave a comment

Comment (jeffhostetler) [inline compat/vcbuild/scripts/clink.pl]: Passing thru the various "-Z" arguments allows caller to specify PDB options.

File: Makefile (hunk 13 above)
Comment (jeffhostetler) [inline Makefile]: I made some simplifications in this morning's commit, but various PDB patterns aren't as straight-forward as the EXEs because some of the EXEs are just copies of the others and don't have their own PDB.  Granted, I could write a pattern and filter out the non-existent ones, but I haven't done that yet.

**Comment by dscho** [comment]
I rebased this branch (verifying that the OpenSSL SHA-1 patch is no longer necessary) and touched up the commit messages and pushed the result to the `vs2015` branch of https://github.com/dscho/git.

I will continue tomorrow, editing the commits as per my suggestions, and hope you may find time some time later this week to look over the branch and maybe rebase your branch?

**Comment by dscho** [comment]
For the record, I've been continuing to work on this branch, and as of an hour ago, it passes the test suite. This required adding the libiconv dependency and a very ugly workaround for the NuPkg's `iconv()` apparently forgetting to set `errno`, but hey,: it passes!

Currently I am testing whether things work just fine without the ugly `SNPRINTF_RETURNS_BOGUS` workaround, and I will still want to check back with you whether the explanation in the other `WIP` commit is correct, but we're getting closer!

**Comment by jeffhostetler** [comment]
Excellent!!!  Thanks for your help on this.
I'll take a look at the other WIP question.
I assume you mean:
    https://github.com/dscho/git/commit/ec7f740032c2f163d86f74671a03302c8436cd0e

**Comment by jeffhostetler** [comment]
I just built your branch and commented out the #ifdef's in https://github.com/dscho/git/commit/ec7f740032c2f163d86f74671a03302c8436cd0e 
and it does build fine.  It reports a couple of warnings, but there are other files where we get those, so if you're OK with a few warnings, we could omit this commit.  (I'm using a pre-release of VS "15" if that matters.)

        CC compat/msvc.o
    msvc.c
    d:\w\gfw\gh\compat\mingw.c(1589): warning C4090: 'function': different 'const' qualifiers
    d:\w\gfw\gh\compat\mingw.c(1599): warning C4090: 'function': different 'const' qualifiers
        AR libgit.a

**Comment by dscho** [comment]
> I'll take a look at the other WIP question.
> I assume you mean:
> dscho@ec7f740

Yep!

> For the record, you're right the commit message is suspect. I should have said that it was complaining about the memcpy() and the free() on the const pointer.

Right! I changed it.

> I just built your branch and commented out the #ifdef's in dscho@ec7f740
> and it does build fine. It reports a couple of warnings, but there are other files where we get those, so if you're OK with a few warnings, we could omit this commit.

I'd rather keep it. It would be nice to eliminate the warnings in the long run (real bugs tend to hide within such swaths of warnings; I remember e.g. an incorrect assignment of `size_t` to `int` that caused spurious segmentation faults on 64-bit systems), but it is not a high priority, I agree.

> (I'm using a pre-release of VS "15" if that matters.)

Don't you have to override `SANE_TOOL_PATH` for that?

**Comment by jeffhostetler** [comment]
WRT VS "15", yes I had to update both SANE_TOOL_PATH and MSVC_VCDIR to get it to build with VS "15".  I haven't pushed this yet, since it's just a pre-release -- and I suspect the lines in config.mak.uname containing "10.0.10240.0" will need a similar fix, but I won't know the build numbers until it is actually released.  Long time, I'd like to have something that properly knows how to get the MSVC paths....

**Comment by dscho** [comment]
> WRT VS "15", yes I had to update both SANE_TOOL_PATH and MSVC_VCDIR to get it to build with VS "15". I haven't pushed this yet, since it's just a pre-release -- and I suspect the lines in config.mak.uname containing "10.0.10240.0" will need a similar fix, but I won't know the build numbers until it is actually released. Long time, I'd like to have something that properly knows how to get the MSVC paths....

I think in the short run, we should allow setting the `SANE_TOOL_PATH` in `config.mak`, i.e. assigning it in `config.mak.uname` with a `?=` ("define unless already defined").

In the longer run, I think we should wait a little and let the new tool layout settle. And once it settled, we can try to figure out a good way to auto-detect the location.

I guess it would be more important, though, to get things building in Visual Studio. The command-line case is really only interesting for unsupervised build jobs.

FWIW my plan is to set up a build job to auto-generate a branch that has the Visual Studio configuration auto-generated from the current `master` branch. That way, we could point interested parties to that branch and they would not even need the Git for Windows SDK to start hacking.

**Comment by jeffhostetler** [comment]
Agreed.  Let's leave it building with 14 and wait the dust to settle.  Thanks again!

**Comment by jeffhostetler** [comment]
@dscho I pulled your version and squashed the 2 PDB commits and pushed it here.

**Comment by dscho** [comment]
Awesome!

I will most likely have to change a few bits here and there when the time comes to try to get this integrated into upstream Git, but for now I am really happy that we have something that builds!

**Comment by jeffhostetler** [comment]
@dscho  Thanks for your help on this!!!

**Comment by Runcy** [comment]
用gcc编译不好吗？

**Comment by dscho** [comment]
> 用gcc编译不好吗？

No, it is not bad to compile with GCC. But Visual C has certain features that we want to use when developing Git for Windows, such as profiling tools, integration with Visual Studio, etc

For future reference: it would be really good to stick with the English language in this project (it is not my first language, either).

**Comment by PhilipOakley** [comment]
> > 用gcc编译不好吗？
[Google translates as] : Gcc compiler do not you?

> No, it is not bad to compile with GCC. But Visual C has certain features that we want to use when developing Git for Windows, such as profiling tools, integration with Visual Studio, etc
  > For future reference: it would be really good to stick with the English language in this project (it is not my first language, either).

  To add to dscho's comments. Yes, the point is to make available tools which are familiar to Windows users (such as me), which include many of the fancy gui / intellisence capabilities that Windows users get to prefer (rightly or wrongly).

  Having multiple tools, such as VS2008 for me, allows me to quickly trace through the code following the call stack, rather than the much more brittle method of going via a local text editor.

  If we can encourage folk to at least have a browse of the code, then they may realise it's not that hard to understand, and hence to contribute.

  Philip
